src/test/java/org/mockitousage/matchers/CapturingArgumentsTest.java

```diff
@@ -298,7 +313,7 @@ public void captures_correctly_when_captor_used_multiple_times() throws Exceptio
 
         // then
         // this is only for backwards compatibility. It does not make sense in real to do so.
-        verify(mock).mixedVarargs(any(), argumentCaptor.capture(), argumentCaptor.capture(), argumentCaptor.capture());
+        verify(mock).mixedVarargs(any(), argumentCaptor.capture());
```

Comment:
@ChristianSchwarz 
Hi,
When you call `capture`, or `captureIf`, the implementation call a `Mockito.argThat`, which stores the matcher in a `ArgumentMatcherStorageImpl`. The code that generates the problem is the following:

```java
    private Set<ArgumentMatcher> uniqueMatcherSet(int indexOfVararg) {
        HashSet<ArgumentMatcher> set = new HashSet<ArgumentMatcher>();
        for (int position = indexOfVararg; position < matchers.size(); position++) {
            ArgumentMatcher matcher = matchers.get(position);
            set.add(matcher);
        }
        return set;
    }
```
Later, this set is used to decide which values are stored in the `arguments` param from `ArgumentCaptor`. On the previous implementation we never had this problem because each time the `capture` method was called, the same instance of `CapturingMatcher` was used (same hashCode, etc.). With the current implementation, each time we call `capture`, we create a new instance of `CapturingMatcher`.
Because of this I decided to implement `equals` and `hashCode` methods, so I don't add to matchers that should capture the same argument (for example, I don't want to add two `isA(String.class)` matchers).

What is the expected result for the following use case:

```java
   ArgumentCaptor<String> argumentCaptor = ArgumentCaptor.forClass(String.class);

        // when
        mock.mixedVarargs(42, "a", "b", "c");

        // then
        // this is only for backwards compatibility. It does not make sense in real to do so.
        verify(mock).mixedVarargs(any(), argumentCaptor.captureIf(isA(String.class)), argumentCaptor.captureIf(isA(String.class)), argumentCaptor.captureIf(isA(String.class)));
```

If we don't implement `equals` and `hashCode`, 9 elements will be captured. Each `isA` matcher will add an element to the `arguments` from `ArgumentCaptor`.
We could decide to keep the old functionality of `capture` and admit the functionality of `captureIf` is the one described above if you use `captureIf` more than once with the same matcher. What do you suggest?